validator/rpc/handlers_health.go

```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		http2.HandleError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ctx.Done():
+			return
+		case <-client.Context().Done():
+			return
+		default:
+			logResp, err := client.Recv()
+			if err != nil {
+				http2.HandleError(w, "could not receive beacon logs from stream", http.StatusInternalServerError)
+				return
+			}
+			jsonResp, err := json.Marshal(logResp)
+			if err != nil {
+				http2.HandleError(w, "could not encode long response", http.StatusInternalServerError)
+				return
+			}
+
+			// Send the response as an SSE event
+			// Assuming resp has a String() method for simplicity
+			_, err = fmt.Fprintf(w, "%s\n", jsonResp)
+			if err != nil {
+				http2.HandleError(w, err.Error(), http.StatusInternalServerError)
+				return
+			}
+			// Flush the data to the client immediately
+			flusher.Flush()
+		}
+	}
+}
+
+// StreamValidatorLogs from the validator client via a gRPC server-side stream.
+func (s *Server) StreamValidatorLogs(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "validator.web.health.StreamValidatorLogs")
+	defer span.End()
+
+	// Ensure that the writer supports flushing.
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		http2.HandleError(w, "Streaming unsupported!", http.StatusInternalServerError)
+		return
+	}
+
+	ch := make(chan []byte, s.streamLogsBufferSize)
+	sub := s.logsStreamer.LogsFeed().Subscribe(ch)
+	defer func() {
+		sub.Unsubscribe()
+		close(ch)
+	}()
+	// Set up SSE response headers
+	w.Header().Set("Content-Type", "text/event-stream")
+	w.Header().Set("Cache-Control", "no-cache")
+	w.Header().Set("Connection", "keep-alive")
+
+	recentLogs := s.logsStreamer.GetLastFewLogs()
+	logStrings := make([]string, len(recentLogs))
+	for i, l := range recentLogs {
+		logStrings[i] = string(l)
+	}
+	ls := &pb.LogsResponse{
+		Logs: logStrings,
+	}
+	jsonLogs, err := json.Marshal(ls)
+	if err != nil {
+		http2.HandleError(w, "Failed to marshal logs", http.StatusInternalServerError)
+		return
+	}
+	_, err = fmt.Fprintf(w, "%s\n", jsonLogs)
+	if err != nil {
+		http2.HandleError(w, "Error sending data", http.StatusInternalServerError)
+		return
+	}
+	flusher.Flush()
+
+	for {
+		select {
+		case log := <-ch:
+			// Set up SSE response headers
+			ls = &pb.LogsResponse{
+				Logs: []string{string(log)},
+			}
+			jsonLogs, err = json.Marshal(ls)
+			if err != nil {
+				http.Error(w, "Failed to marshal logs", http.StatusInternalServerError)
+				return
+			}
+			_, err = fmt.Fprintf(w, "%s\n", jsonLogs)
+			if err != nil {
+				http.Error(w, "Error sending data", http.StatusInternalServerError)
+				return
+			}
+
+			flusher.Flush()
+		case <-s.ctx.Done():
+			return
+		case err := <-sub.Err():
+			http.Error(w, fmt.Sprintf("Subscriber error: %v", err), http.StatusInternalServerError)
+			return
+		case <-ctx.Done():
+			return
```

Comment:
We should return an error

Reply:
it complains that you are overriding the header if we do it on context done